File: 1260-shift-2d-grid/1260-shift-2d-grid.java
class Solution {
    public List<List<Integer>> shiftGrid(int[][] grid, int k) {
        int n = grid.length;
        int m = grid[0].length;


        List<List<Integer>> li = new ArrayList<>();
        for(int i = 0; i < n; i++){
            li.add(new ArrayList<>());
            for(int j = 0; j < m; j++){
                li.get(i).add(0);
            }
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                int newPos = (i * m) + j + k;
                int row = newPos / m;
                int col = newPos % m;

                li.get(row % n).set((col % m), grid[i][j]);
            }
        }
        return li;
    }
}

File: 1260-shift-2d-grid/1260-shift-2d-grid.js
/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number[][]}
 */
var shiftGrid = function(grid, k) {
    const n = grid.length;
    const m = grid[0].length;

    k %= (m * n);

    const res = Array(n)
        .fill()
        .map(() => Array(m).fill(0));
    
    for(let i = 0; i < n; i++){
        for(let j = 0; j < m; j++){
            let newPos = (i * m) + j + k;

            let row = Math.floor(newPos / m) % n;
            let col = newPos % m;

            res[row % n][col] = grid[i][j];
        }
    }
    return res;
};